src/operator/semijoin.rs

```diff
@@ -0,0 +1,609 @@
+where
+    Pairs: BatchReader<Time = ()> + 'static,
+    Pairs::Key: Clone + Ord,
+    Pairs::Val: Clone,
+    Pairs::R: HasZero + MulByRef,
+    Keys: BatchReader<Key = Pairs::Key, Val = (), Time = (), R = Pairs::R> + 'static,
+    Out: ZSet<Key = (Pairs::Key, Pairs::Val), R = Pairs::R> + 'static,
+{
+    fn eval(&mut self, pairs: &Pairs, keys: &Keys) -> Out {
+        let mut pair_cursor = pairs.cursor();
+        let mut key_cursor = keys.cursor();
+
+        // Choose capacity heuristically.
+        let mut builder = Out::Builder::with_capacity((), min(pairs.len(), keys.len()));
+
+        // While both keys are valid
+        // TODO: Is there a better way to iterate here? `keys_cursor` is the
+        //       thing really driving this, so can we just use it as the
+        //       source of iteration to do the least work possible?
+        while key_cursor.key_valid() && pair_cursor.key_valid() {
+            match key_cursor.key().cmp(pair_cursor.key()) {
+                // Match up both the cursors
+                Ordering::Less => key_cursor.seek_key(pair_cursor.key()),
+                Ordering::Greater => pair_cursor.seek_key(key_cursor.key()),
+
+                Ordering::Equal => {
+                    // TODO: Can the value of `()` ever be invalid? Do we need an `if
+                    // key_cursor.val_valid()` check?
+                    let key_weight = key_cursor.weight();
+                    while pair_cursor.val_valid() {
+                        // Get the weight of the output kv pair by multiplying them together
+                        // TODO: Can either weights possibly be zero? If so, we can check if
+                        // `key_weight`       is zero outside of the loop to
+                        // skip redundant work
+                        let pair_weight = pair_cursor.weight();
+                        let kv_weight = pair_weight.mul_by_ref(&key_weight);
+
+                        // Add to our output batch
+                        builder.push((
+                            (pair_cursor.key().clone(), pair_cursor.val().clone()),
+                            (),
+                            kv_weight,
+                        ));
+                        pair_cursor.step_val();
+                    }
+
+                    pair_cursor.step_key();
+                    key_cursor.step_key();
+                }
+            }
+        }
+
+        // Create the output stream
+        builder.done()
+    }
+}
+
+/// Semijoin two streams of batches, see [`Stream::semijoin_stream_core`]
+pub struct SemiJoinStream<F, Pairs, Keys, Out> {
+    semijoin_func: F,
+    _types: PhantomData<(Pairs, Keys, Out)>,
+}
+
+impl<F, Pairs, Keys, Out> SemiJoinStream<F, Pairs, Keys, Out> {
+    pub const fn new(semijoin_func: F) -> Self {
+        Self {
+            semijoin_func,
+            _types: PhantomData,
+        }
+    }
+}
+
+impl<F, Pairs, Keys, Out> Operator for SemiJoinStream<F, Pairs, Keys, Out>
```

Comment:
Why do we need the non-monotonic implementation?  Is there ever a situation when the monotonic one doesn't work?  Besides, this is equivalent (both functionality and perforance-wise) to `stream_join`, isn't it?

Reply:
The non-monotonic implementation uses a `Builder` instead of a `Batcher`, meaning it's potentially more efficent

Reply:
You mean the _monotonic_ one? I understand why we need that. But the non-monotonic uses batches, so should be identical to normal join, shouldn't it?